### PR TITLE
fix: torch.compile(fullgraph=True) compatibility for norm ops

### PR DESCRIPTION
**Pull request by Johnsonms**
## 📌 Description
  https://github.com/flashinfer-ai/flashinfer/issues/2733

`register_custom_op` / `register_fake_op` in `utils.py` were no-ops, causing JIT-backed FlashInfer ops to behave like plain Python functions. Under `torch.compile(fullgraph=True)`, Dynamo would trace into the Python implementation and hit `Path.exists() -> posix.stat`, which it cannot trace, breaking compilation.

  This PR re-enables `torch.library.custom_op` and
  `torch.library.register_fake` so Dynamo treats these ops as opaque custom ops
  instead of entering the Python body.

  It also removes the now-unnecessary `@torch.compiler.disable` from
  `get_norm_module` in `norm.py`, and adds
  `tests/norm/test_norm_torch_compile.py` with 40 regression cases covering:

  - `flashinfer.rmsnorm`
  - `flashinfer.fused_add_rmsnorm`
  - `flashinfer.gemma_rmsnorm`
  - `flashinfer.gemma_fused_add_rmsnorm`

  all under `torch.compile(fullgraph=True)`.

  Scope note: the regression coverage added in this PR is for norm ops.

  ## 🔍 Related Issues

  Fixes #<issue_number>

  ## 🚀 Pull Request Checklist

  Thank you for contributing to FlashInfer! Before we review your pull request, please make sure the following items are complete.

  ### ✅ Pre-commit Checks

  - [x] I have installed `pre-commit` by running `pip install pre-commit` (or used your preferred method).
  - [x] I have installed the hooks with `pre-commit install`.
  - [x] I have run the hooks manually with `pre-commit run --all-files` and fixed any reported issues.

  > If you are unsure about how to set up `pre-commit`, see [the pre-commit documentation](https://pre-commit.com/).

  ## 🧪 Tests

  - [x] Tests have been added or updated as needed.
  - [x] All tests are passing (`unittest`, etc.).
    - `pytest tests/norm/test_norm_torch_compile.py -v`
    - `40 passed`

  ## Reviewer Notes

  This PR fixes the `torch.compile(fullgraph=True)` regression for the norm path.
  The custom-op registration change is shared infrastructure, but the added
  regression coverage in this PR is norm-only.

## Release Notes

* **Tests**
  * Added comprehensive torch.compile compatibility tests for multiple RMSNorm variants across various batch sizes, hidden dimensions, and data types (float16, bfloat16), validating compiled execution against reference results.

* **Improvements**
  * Replaced previous stub behavior with direct PyTorch library integration for custom/fake operation registration, improving runtime alignment with PyTorch.

### CODE REVIEW

File: tests/norm/test_norm_torch_compile.py [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+def test_gemma_rmsnorm_torch_compile(batch_size, hidden_size, dtype):
+    x = torch.randn(batch_size, hidden_size, dtype=dtype, device="cuda")
+    w = torch.zeros(hidden_size, dtype=dtype, device="cuda")  # gemma: (w+1)*norm
+
+    # warmup
+    flashinfer.gemma_rmsnorm(x, w)
+
+    compiled = torch.compile(flashinfer.gemma_rmsnorm, fullgraph=True, backend="eager")
+    out = compiled(x, w)
+    ref = flashinfer.gemma_rmsnorm(x, w)
```
[the remark this answers is not in the document]
Comment (Johnsonms) [reply tests/norm/test_norm_torch_compile.py]: fixed

File: tests/norm/test_norm_torch_compile.py [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+
+Verifies that rmsnorm / fused_add_rmsnorm / gemma_rmsnorm /
+gemma_fused_add_rmsnorm can be traced by Dynamo with fullgraph=True
+(i.e. no graph breaks, no posix.stat issues).
+"""
+
+import pytest
+import torch
+
+import flashinfer
```
[the remark this answers is not in the document]
Comment (Johnsonms) [reply tests/norm/test_norm_torch_compile.py:30]: Fixed

File: flashinfer/utils.py [hunk 3]
```diff
@@ -344,24 +344,19 @@ def register_custom_op(
...
-
-        # return torch.library.custom_op(
-        #     name,
-        #     fn,
-        #     mutates_args=mutates_args,
-        #     device_types=device_types,
-        #     schema=schema,
-        # )
-        return lambda x: x
+        return torch.library.custom_op(
```
Comment (yzh119) [inline flashinfer/utils.py:347]: If we turn this on again, it's more than just normalization kernel and we probably need to start working on cudagraph sanity check across the codebase (but let's see how CI works first).

As far as I remember sglang wraps these API again with torch library decorators so it should not be a problem?

Comment (nvpohanh) [reply flashinfer/utils.py:347]: @Johnsonms is it possible for SGLang to wrap the FlashInfer op on SGLang-side?

Comment (Johnsonms) [reply flashinfer/utils.py:347]: Thanks @nvpohanh @yzh119 !

Yes @yzh119, you are right — SGLang (sgl-kernel) already wraps these ops. For flexibility, we currently support JIT kernels, where the SGLang wrapper directly calls the FlashInfer JIT implementation.

The issue with torch.compile(..., fullgraph=True) is not just the wrapper itself. Dynamo still sees the underlying FlashInfer JIT-backed APIs as regular Python functions, so it traces into them.

During tracing it eventually hits the JIT loading logic (e.g., Path.exists() / os.stat()), which Dynamo cannot capture. As a result, CUDA graph–wise compilation breaks or fails in this path.

Comment (nvpohanh) [reply flashinfer/utils.py:347]: @yzh119 Is it possible that in FlashInfer, we somehow "detect" if it is in capture mode and if so, we skip those JIT loading logic?

Comment (nvjullin) [reply flashinfer/utils.py:347]: @Johnsonms I'm a bit confused.
If sglang wraps with custom op, why does dynamo trace into the op?
If not, why not?

Comment (Johnsonms) [reply flashinfer/utils.py:347]: Good question @nvjullin , Let me take time to investigate and get the everything clear tomorrow

Comment (Johnsonms) [reply flashinfer/utils.py:347]: There is an online urgent issue, having to work there, will do this in weekend. Sorry about that

Comment (DevashishLal-CB) [reply flashinfer/utils.py:347]: @Johnsonms @yzh119 @nvpohanh I am using these kernels in sglang for fusion by registering them on sglang side

https://github.com/sgl-project/sglang/pull/10549/changes#diff-a3005a66a8755c5a5c2043c67c32bf0be63134bd45b4cd70455ada4c06cb3d0fR51

It does work but it would be nice to enable this support in flashinfer to avoid this duplication for all the kernels I want to use.

not sure if performance is still a concern but vllm and sglang work around this by using the direct registration of custom ops

https://github.com/vllm-project/vllm/blob/36e76700453924c8d421db99af70a88a1df835cd/vllm/utils.py#L1696

Comment (Johnsonms) [reply flashinfer/utils.py]: Thanks @nvjullin — I looked into this and the root cause is now clearer.

The FlashInfer JIT kernel APIs (e.g. `flashinfer.norm.rmsnorm`) currently appear as plain Python functions to Dynamo. Although FlashInfer has `register_custom_op` / `register_fake_op` decorators, those registrations are effectively no-ops right now, so the functions are not exposed as real PyTorch custom ops.

On the SGLang side we call these FlashInfer Python APIs directly. Since Dynamo does not see an opaque custom op boundary, it traces into the Python implementation. During tracing it eventually reaches the JIT loading logic (`build_and_load()`, `Path.exists()`, `os.stat()`), which Dynamo cannot capture, and this breaks `torch.compile(..., fullgraph=True)`.

This is also why the internal `torch.ops.sgl_kernel.*` (or we called AOT) path works: those kernels are properly registered as PyTorch operators in C++, so Dynamo treats them as opaque ops and does not trace into them.

The current workaround (as @DevashishLal-CB  mentioned) is to register the kernels as custom ops on the SGLang side. A cleaner fix would be for FlashInfer to expose these JIT-backed APIs as proper torch custom ops so Dynamo treats them as opaque operators.

Comment (nvjullin) [reply flashinfer/utils.py:347]: Is this a correct summary then?
1. `torch.compile` works after wrapping flashinfer functions in custom ops
2. But sglang would like the custom op wrapping be in flashinfer instead

Comment (nvpohanh) [reply flashinfer/utils.py:347]: @xinli-sw could you find someone from vLLM team to review this? I think the original perf concern was raised by vLLM team

Comment (benchislett) [reply flashinfer/utils.py:347]: I assume vLLM will want to wrap the custom op itself, but not sure about the performance implications here.

Comment (benchislett) [reply flashinfer/utils.py:347]: I have seen the `posix.stat` crash before so we will want some kind of workaround for this issue. I don't have a strong intuition for whether FI should be the one to wrap the op or not.

Comment (nvjullin) [reply flashinfer/utils.py:347]: > not sure about the performance implications here

I believe the history was vLLM asked flashinfer to turn **off** wrapping, because they wanted to hack into the torch custom op machinery https://github.com/vllm-project/vllm/pull/9896. The main motivation being perf.
